Web Inspector: Styles Panel: Adding a new CSS rule doesn't work on first attempt

rcaliman-apple opened the pull request:
#### 96c03ed0c66bcdd31e22b6e6d81ebaebbd22e47d
<pre>
Web Inspector: Styles Panel: Adding a new CSS rule doesn&apos;t work on first attempt
<a href="https://bugs.webkit.org/show_bug.cgi?id=253956">https://bugs.webkit.org/show_bug.cgi?id=253956</a>

Reviewed by Patrick Angle.

To add a new CSS rule, an inspector stylesheet is first created.
This operation triggers a `DOMNodeStyles.refresh()`.

The operation to add the new rule to this stylesheet also triggers a
refresh. But if the promise for the first refresh is still pending,
the second call is ignored. This means that the Styles panel ends up
not showing any matching styles from the initially empty stylehseet.

A subsequent call to add a new rule bypasses the need to create a new
stylesheet. The `DOMNodeStyles.refresh()` it trigges brings the Styles
panel in sync with the latest contents of the inspector-generated stylehseet.
That&apos;s why two rules show up on the first successful attempt.

The logic to mark the newly created rule&apos;s selector as editable is too lenient:
it marks all CSS rules with a selector identical to the one of the newly created rule.
This patch makes a stricter check by identifying the newly created rule by its `CSSStyle.styleId`.

* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles.prototype.addRule.completed):
(WI.DOMNodeStyles.prototype.addRule.addedRule):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
(WI.SpreadsheetRulesStyleDetailsPanel):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype.spreadsheetCSSStyleDeclarationSectionAddNewRule):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype.layout):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype._addNewRule):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype._addedNewRuleCallback):

Canonical link: <a href="https://commits.webkit.org/261883@main">https://commits.webkit.org/261883@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03e82ca796279f932f353d6e18573c57efb5b555

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22026 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1544 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4648 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121367 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23373 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13193 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5827 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118629 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17362 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100621 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105957 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99321 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1158 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46379 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14324 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1197 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/98817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15026 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10533 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20341 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53190 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8299 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16876 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -367,7 +367,15 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
+            // then refresh again to get the latest matching styles which include the newly created rule.
+            if (this._pendingRefreshTask) {
+                this._pendingRefreshTask.then(this.refresh.bind(this));
+            }
```
Style: single-line conditional bodies don't need enclosing `{}`

patrickangle commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -358,7 +358,7 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
     }
 
-    addRule(selector, text, styleSheetId)
+    addRule(selector, text, styleSheetId, addedRuleCallback)
```
It seems like we should either make this method `async` and have it return the new new rule ID

rcaliman-apple replied:
I don't think that'll work.

The entire chain of operations within `DOMNodeStyles.addRule()` ends with a call to `DOMNodeStyles.refresh()`. That will trigger a `WI.DOMNodeStyles.Event.Refreshed` event in response to which `WI.SpreadsheetRulesStyleDetailsPanel` will do layout. The `styleID` is needed in that layout step in order to mark the correct rule as editable.

Awaiting the full set of operations will lead to a race condition.

Here we're intentionally calling the `addedRuleCalback` synchronously as soon as the rule is created, before any other method has a chance to call `DOMNodeStyles.refresh()` and trigger the layout above.

patrickangle replied:
Since `refresh` returns a promise (either the existing refresh promise or a new refresh promise), it seems we should just be able to call `refresh` but not await the promise it returns, effectively scheduling the refresh for later, right? Effectively I'm suggesting that addRule schedule that work and only wait for the new rule's styleID to exist before fulfilling its promise.

In practice this would only require you to do something like `let result = new WI.WrappedPromise;` at the start of this function, return/await `result.promise` at the end of the function, and instead of calling the callback you'd call `result.resolve(rulePayload);`.

This has a few benefits. It allows the caller to be asynchronous in the future if need be without having to wrap this function in special logic. It has an added benefit of throwing an error if you try to resolve the wrapped promise more than once. And it benefits from more closely matching the semantics of how `changeRuleSelector` behaves, returning a promise with the resulting rule payload.

rcaliman-apple replied:
Oh, neat! That'll work. Updated the patch.

patrickangle reviewed (commented):
You lost me in your explanation in the commit message. You went straight from describing the problem to describing what this patch does, but it's unclear to me why this fixes the problem from the commit message. Your comment on DOMNodeStyles.js:371 is likely the missing piece there, but it seems wrong that we have to wait for the pending refresh just to refresh again... that said, I don't have a better suggestion at this time.

rcaliman-apple committed:
Web Inspector: Styles Panel: Adding a new CSS rule doesn't work on first attempt
https://bugs.webkit.org/show_bug.cgi?id=253956

Reviewed by Patrick Angle.

To add a new CSS rule, an inspector stylesheet is first created.
This operation triggers a `DOMNodeStyles.refresh()`.

The operation to add the new rule to this stylesheet also triggers a
refresh. But if the promise for the first refresh is still pending,
the second call is ignored. This means that the Styles panel ends up
not showing any matching styles from the initially empty stylehseet.

A subsequent call to add a new rule bypasses the need to create a new
stylesheet. The `DOMNodeStyles.refresh()` it trigges brings the Styles
panel in sync with the latest contents of the inspector-generated stylehseet.
That's why two rules show up on the first successful attempt.

The logic to mark the newly created rule's selector as editable is too lenient:
it marks all CSS rules with a selector identical to the one of the newly created rule.
This patch makes a stricter check by identifying the newly created rule by its `CSSStyle.styleId`.

* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles.prototype.addRule.completed):
(WI.DOMNodeStyles.prototype.addRule.addedRule):
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRulesStyleDetailsPanel.js:
(WI.SpreadsheetRulesStyleDetailsPanel):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype.spreadsheetCSSStyleDeclarationSectionAddNewRule):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype.layout):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype._addNewRule):
(WI.SpreadsheetRulesStyleDetailsPanel.prototype._addedNewRuleCallback):

Canonical link: https://commits.webkit.org/261883@main
